dbms/src/IO/Compression/CompressionCodecMultiple.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/PODArray.h>
+#include <IO/Compression/CompressionCodecMultiple.h>
+#include <IO/Compression/CompressionFactory.h>
+#include <IO/Compression/CompressionInfo.h>
+#include <IO/Operators.h>
+#include <IO/WriteHelpers.h>
+#include <common/unaligned.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int CORRUPTED_DATA;
+}
+
+CompressionCodecMultiple::CompressionCodecMultiple(Codecs codecs_)
+    : codecs(codecs_)
+{}
+
+UInt8 CompressionCodecMultiple::getMethodByte() const
+{
+    return static_cast<UInt8>(CompressionMethodByte::Multiple);
+}
+
+UInt32 CompressionCodecMultiple::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    UInt32 compressed_size = uncompressed_size;
+    for (const auto & codec : codecs)
+        compressed_size = codec->getCompressedReserveSize(compressed_size);
+
+    ///    TotalCodecs  ByteForEachCodec       data
+    return static_cast<UInt32>(sizeof(UInt8) + codecs.size() + compressed_size);
+}
+
+UInt32 CompressionCodecMultiple::doCompressData(const char * source, UInt32 source_size, char * dest) const
+{
+    PODArray<char> compressed_buf;
+    PODArray<char> uncompressed_buf(source, source + source_size);
+
+    dest[0] = static_cast<UInt8>(codecs.size());
+
+    size_t codecs_byte_pos = 1;
```

Comment:
Seems `codecs_byte_pos` is always `idx` + 1?

Reply:
addressed